libmscore/hairpin.cpp

```diff
@@ -108,11 +108,13 @@ void HairpinSegment::layout()
                         // systems may be unknown at layout stage.
                         ed = toDynamic(end->findAnnotation(ElementType::DYNAMIC, _track, _track));
                         }
-                  if (ed) {
+                  if (ed && ed->visible() && ed->autoplace()) {
                         const qreal edLeft  = ed->bbox().left() + ed->pos().x()
                                               + ed->segment()->pos().x() + ed->measure()->pos().x();
                         const qreal dist    = edLeft - pos2().x() - pos().x() - score()->styleP(Sid::autoplaceHairpinDynamicsDistance);
                         rxpos2() += dist;
+                        // TODO - don't extend hairpin across barline to reach dynamic in next measure?
+                        // or only if there is also a key signature? see Gould
```

Comment:
Another relevant question: does it necessary to do horizontal shifting of dynamic marks? The original code I restored included that (see #4203) but I stripped it down to make a simple solution to avoid a completely bad layout in such cases — the real aligning is available as begin/end text for hairpins anyway. Still with some accuracy and a slight reordering of elements layout this can be returned back, even in a simpler way than it was before. Although that is not an urgent issue anyway, it would be good to know whether we need to think about it.

Reply:
I am afraid I am not quite understanding what you are asking here.  Are you taking about shifting the dynamic markings themselves,, as opposed to the hairpin?  I think not by default, although subjectively there might be places it makes sense for someone to do manually.  Maybe have a minimum hairpin length, and shift dynamics if the adjustment here shortens the hairpin below that threshold?

As for the "real aligning" you mention - to me, that's not much of an option if it doesn't affect playback, and you can't get the actual dynamic glyphs in there.  Maybe I'm missing something about how that works?

Reply:
Yes, I asked about shifting both dynamic markings and hairpin, like it was done by the old code for that. Though dynamics shifting was probably not too big.

Concerning placing dynamic glyphs as text, actually you can currently do that, for example, by dragging a dynamic marking from palette directly to the hairpin rather than note. Advanced workspace has also a predefined "dynamics + hairpin" element.
Playback indeed seems to be not implemented at the moment, but as far as I know @wschweer has some plans to implement it, if I am correct.

Reply:
Thanks, I hadn't realized drag & drop worked that way.  Although I don't see a way to get the dynamic at the *beginning* that way.  No big deal, that's not the more important use case.

But actually, it doesn't quite address the case I was concerned with.  Here are three examples.  

1) First, what I really want to see:

![crescf-1](https://user-images.githubusercontent.com/1799936/49700135-1292a000-fb98-11e8-8fdd-7d6fda729b4a.png)

2) Here's what I get by default using the "regular" way of adding dynamics:

![crescf-2](https://user-images.githubusercontent.com/1799936/49700138-20e0bc00-fb98-11e8-8824-37908a2cb913.png)

3) Here's what I get using the "drop directly on hairpin" method:

![crescf-3](https://user-images.githubusercontent.com/1799936/49700141-29d18d80-fb98-11e8-86fe-7b18493c1741.png)

What I propose is automatically turning 2) into 1).  I got what I wanted by disabling autoplace.  But since this is something of a personal preference (and as suggested in the comment, Gould gives leeway here), I think a style parameter makes some sense.  That is, we calculate the amount by which we would need to extending the hairpin, and onyl do it if it exceeds some threshold.  So basically, the presence of the key signature, or clef, or lyric, or whatever, that pushes the first note of the measure to the right would end up being the determing factor.

Reply:
The primary intention of returning that code was anyway to shorten hairpins if dynamics marking is wide enough to push that hairpin down. So we could even disable elongation of hairpins at all, or indeed make some threshold for it if appropriate.